# __API REST__

Un API REST es una interfaz web que permite permite realizar operaciones CRUD utilizando peticiones HTTP:
* Create: POST
* Update: PUT, PATCH
* Read: GET
* Delete: DELETE

Normalmente el intercambio de información se produce mediante documentos JSON o XML (menos común). Las API REST **son interfaces sin estado**, es decir, cada petición es independiente de las demás y por ello deben incluir todos los datos necesarios para procesarla. Es por ello que la autenticación se realiza mediante _tokens_ de acceso que se deben obtener previamente y que se incluyen en cada una de las peticiones.

## Ejemplo con OMDb API

Se va a ver un ejemplo de API-REST en la que se usará la función GET para recuperar recursos que tienen asociado un identificador único  en forma de URI. La  URI está formada por la dirección del recurso y un diccionario con las opciones de búsqueda con claves que representan los parámetros debúsqueda y los valores que deben tomar.

El ejemplo que se va a usar un portal de datos abiertos sobre películas y series denominado OMDb API que se encuentra en:

http://www.omdbapi.com/

Para usarla hay que obtener una clave de uso en:

http://www.omdbapi.com/apikey.aspx

![Obtenición del token](APIkey_OMDB.png)

En la página se encuentra una explicación de cómo usar la API:

![Parémetros](OMDB_params.png)

Se va crear un ejemplo para recuperar información de la película "Guardians of the Galaxy". Se usa el parámetro s que representa el título de la película y el parámero type para indicar que es una película.

In [1]:
# Pasos previos:
# 1) Crear un token en http://www.omdbapi.com/apikey.aspx (a veces tarda un rato en estar operativo)
# 2) Copiar el token de acceso en el fichero 'token_omdb.txt' en el mismo directorio del notebook

with open('token_omdb.txt', 'r', encoding='utf8') as f:
    TOKEN_OMDB = f.read().strip()
    # print(TOKEN_OMDB)

In [2]:
import requests

URL = "http://www.omdbapi.com/"

response = requests.get( URL,
    params={'apikey': TOKEN_OMDB,
            's': 'Guardians of the Galaxy', 
            'type': 'movie',
            'r': 'json'
           },
)
results = response.json()
results

{'Search': [{'Title': 'Guardians of the Galaxy',
   'Year': '2014',
   'imdbID': 'tt2015381',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BNDIzMTk4NDYtMjg5OS00ZGI0LWJhZDYtMzdmZGY1YWU5ZGNkXkEyXkFqcGdeQXVyMTI5NzUyMTIz._V1_SX300.jpg'},
  {'Title': 'Guardians of the Galaxy Vol. 2',
   'Year': '2017',
   'imdbID': 'tt3896198',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BNjM0NTc0NzItM2FlYS00YzEwLWE0YmUtNTA2ZWIzODc2OTgxXkEyXkFqcGdeQXVyNTgwNzIyNzg@._V1_SX300.jpg'},
  {'Title': 'Guardians of the Galaxy Vol. 3',
   'Year': '2023',
   'imdbID': 'tt6791350',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BMDgxOTdjMzYtZGQxMS00ZTAzLWI4Y2UtMTQzN2VlYjYyZWRiXkEyXkFqcGdeQXVyMTkxNjUyNQ@@._V1_SX300.jpg'},
  {'Title': 'The Guardians of the Galaxy Holiday Special',
   'Year': '2022',
   'imdbID': 'tt13623136',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BZDA3MzdlYTQtMTUxNi00ZjJmLTkyOTYtNDkzYmIzY

In [3]:
import requests

URL = "http://www.omdbapi.com/"

response = requests.get( URL,
    params={'apikey': TOKEN_OMDB,
            's': 'Guardians of the Galaxy', 
            'type': 'movie',
            'r': 'xml'
           },
)
results = response.text
results

'<root totalResults="21" response="True"><result title="Guardians of the Galaxy" year="2014" imdbID="tt2015381" type="movie" poster="https://m.media-amazon.com/images/M/MV5BNDIzMTk4NDYtMjg5OS00ZGI0LWJhZDYtMzdmZGY1YWU5ZGNkXkEyXkFqcGdeQXVyMTI5NzUyMTIz._V1_SX300.jpg"/><result title="Guardians of the Galaxy Vol. 2" year="2017" imdbID="tt3896198" type="movie" poster="https://m.media-amazon.com/images/M/MV5BNjM0NTc0NzItM2FlYS00YzEwLWE0YmUtNTA2ZWIzODc2OTgxXkEyXkFqcGdeQXVyNTgwNzIyNzg@._V1_SX300.jpg"/><result title="Guardians of the Galaxy Vol. 3" year="2023" imdbID="tt6791350" type="movie" poster="https://m.media-amazon.com/images/M/MV5BMDgxOTdjMzYtZGQxMS00ZTAzLWI4Y2UtMTQzN2VlYjYyZWRiXkEyXkFqcGdeQXVyMTkxNjUyNQ@@._V1_SX300.jpg"/><result title="The Guardians of the Galaxy Holiday Special" year="2022" imdbID="tt13623136" type="movie" poster="https://m.media-amazon.com/images/M/MV5BZDA3MzdlYTQtMTUxNi00ZjJmLTkyOTYtNDkzYmIzYTJkZjMzXkEyXkFqcGc@._V1_SX300.jpg"/><result title="Guardians of the Galaxy: 

Ahora se puede tomar el identificador (campo `imdbID`) de la película para obtener más información de la misma. Concretamente sobre la película de `imdbID=tt2015381`, que es 'Guardians of the Galaxy Vol. 2'.

In [4]:
response = requests.get(URL,
    params={'apikey': TOKEN_OMDB,
            'i': 'tt3896198',
            'plot': 'full'}
)
results = response.json()
results

{'Title': 'Guardians of the Galaxy Vol. 2',
 'Year': '2017',
 'Rated': 'PG-13',
 'Released': '05 May 2017',
 'Runtime': '136 min',
 'Genre': 'Action, Adventure, Comedy',
 'Director': 'James Gunn',
 'Writer': 'James Gunn, Dan Abnett, Andy Lanning',
 'Actors': 'Chris Pratt, Zoe Saldana, Dave Bautista',
 'Plot': "After saving Xandar from Ronan's wrath, the Guardians are now recognized as heroes. Now the team must help their leader Star Lord (Chris Pratt) uncover the truth behind his true heritage. Along the way, old foes turn to allies and betrayal is blooming. And the Guardians find that they are up against a devastating new menace who is out to rule the galaxy.",
 'Language': 'English',
 'Country': 'United States',
 'Awards': 'Nominated for 1 Oscar. 15 wins & 60 nominations total',
 'Poster': 'https://m.media-amazon.com/images/M/MV5BNjM0NTc0NzItM2FlYS00YzEwLWE0YmUtNTA2ZWIzODc2OTgxXkEyXkFqcGdeQXVyNTgwNzIyNzg@._V1_SX300.jpg',
 'Ratings': [{'Source': 'Internet Movie Database', 'Value': '7.

Ahora vamos a buscar todas las series que contienen `Lord of the Rings` en su título para ordenarlas de mayor a menor `imdbRating`:

In [5]:
response = requests.get( URL,
    params={'apikey': TOKEN_OMDB,
            's': 'lord of the rings',   # OJO: no busca frase exacta
            'type': 'movie',
            'page': 1
           },
)
results = response.json()
results

{'Search': [{'Title': 'The Lord of the Rings: The Fellowship of the Ring',
   'Year': '2001',
   'imdbID': 'tt0120737',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BNzIxMDQ2YTctNDY4MC00ZTRhLTk4ODQtMTVlOWY4NTdiYmMwXkEyXkFqcGc@._V1_SX300.jpg'},
  {'Title': 'The Lord of the Rings: The Return of the King',
   'Year': '2003',
   'imdbID': 'tt0167260',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BMTZkMjBjNWMtZGI5OC00MGU0LTk4ZTItODg2NWM3NTVmNWQ4XkEyXkFqcGc@._V1_SX300.jpg'},
  {'Title': 'The Lord of the Rings: The Two Towers',
   'Year': '2002',
   'imdbID': 'tt0167261',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BMGQxMDdiOWUtYjc1Ni00YzM1LWE2NjMtZTg3Y2JkMjEzMTJjXkEyXkFqcGc@._V1_SX300.jpg'},
  {'Title': 'The Lord of the Rings',
   'Year': '1978',
   'imdbID': 'tt0077869',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BZmI4ZmIxOGQtMGY2ZS00Y2Y5LTllMDItYzllOWFmMTNlMmY2XkEyXkFqcGc@._V1

In [6]:
# Detectar el número de páginas a leer, usando paginación

RESULTS_PER_PAGE = 10

nres = int(results['totalResults'])
print(len(results['Search']))
npages = nres // RESULTS_PER_PAGE
npages = npages + 1 if nres % RESULTS_PER_PAGE > 0 else npages
npages

10


5

In [7]:
from pprint import pp

# Lista de todas las páginas como diccionarios
pages = [requests.get(URL, params={'apikey': TOKEN_OMDB,
                                   's': 'lord of the rings',
                                   'type': 'movie',
                                   'page': i}).json().get('Search') for i in range(1, npages+1)]
pp(pages)

# Lista de los imdbID de películas que **de verdad contienen 'lord of the rings'**
ids = [movie['imdbID'] for page in pages for movie in page if 'lord of the rings' in movie['Title'].lower()]
pp(ids)
print(len(ids))

[[{'Title': 'The Lord of the Rings: The Fellowship of the Ring',
   'Year': '2001',
   'imdbID': 'tt0120737',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BNzIxMDQ2YTctNDY4MC00ZTRhLTk4ODQtMTVlOWY4NTdiYmMwXkEyXkFqcGc@._V1_SX300.jpg'},
  {'Title': 'The Lord of the Rings: The Return of the King',
   'Year': '2003',
   'imdbID': 'tt0167260',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BMTZkMjBjNWMtZGI5OC00MGU0LTk4ZTItODg2NWM3NTVmNWQ4XkEyXkFqcGc@._V1_SX300.jpg'},
  {'Title': 'The Lord of the Rings: The Two Towers',
   'Year': '2002',
   'imdbID': 'tt0167261',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BMGQxMDdiOWUtYjc1Ni00YzM1LWE2NjMtZTg3Y2JkMjEzMTJjXkEyXkFqcGc@._V1_SX300.jpg'},
  {'Title': 'The Lord of the Rings',
   'Year': '1978',
   'imdbID': 'tt0077869',
   'Type': 'movie',
   'Poster': 'https://m.media-amazon.com/images/M/MV5BZmI4ZmIxOGQtMGY2ZS00Y2Y5LTllMDItYzllOWFmMTNlMmY2XkEyXkFqcGc@._V1_SX300.jpg

In [8]:
# Lista de (puntuacion, titulo, año) de las películas a partir de sus ids
# Para no saturar demasiado el servidor, pediremos únicamente las 15 primeras cuyo imdbRating != 'N/A'

info_peliculas = [(float(d['imdbRating']), d['Title'], d['Year'])
                  for imdbid in ids[:15] 
                  if (d := requests.get(URL, params={'apikey': TOKEN_OMDB, 'i': imdbid}).json()) is not None
                      and d['imdbRating'] != 'N/A'
                 ]
pp(info_peliculas)
print(len(info_peliculas))

[(8.9, 'The Lord of the Rings: The Fellowship of the Ring', '2001'),
 (9.0, 'The Lord of the Rings: The Return of the King', '2003'),
 (8.8, 'The Lord of the Rings: The Two Towers', '2002'),
 (6.2, 'The Lord of the Rings', '1978'),
 (7.8, "The Making of 'The Lord of the Rings'", '2002'),
 (7.6, "A Passage to Middle-earth: The Making of 'Lord of the Rings'", '2001'),
 (6.3,
  'National Geographic: Beyond the Movie - The Lord of the Rings: Return of '
  'the King',
  '2003'),
 (6.6, 'The Lord of the Rings: The Quest Fulfilled', '2003'),
 (6.6,
  "Master of the Rings: The Unauthorized Story Behind J.R.R. Tolkien's 'Lord "
  "of the Rings'",
  '2001'),
 (7.9,
  "Creating the Lord of the Rings Symphony: A Composer's Journey Through "
  'Middle-Earth',
  '2004'),
 (8.9,
  'The Lord of the Rings - The Appendices Part 1: From Book to Vision',
  '2002'),
 (6.9, "Tolkien's the Lord of the Rings: A Catholic Worldview", '2011'),
 (8.9, 'The Lord of the Rings Trilogy: Behind-the-Scenes', '2006'),
 

In [9]:
sorted(info_peliculas, reverse=True)

[(9.0, 'The Lord of the Rings: The Return of the King', '2003'),
 (8.9, 'The Lord of the Rings: The Fellowship of the Ring', '2001'),
 (8.9, 'The Lord of the Rings Trilogy: Behind-the-Scenes', '2006'),
 (8.9,
  'The Lord of the Rings - The Appendices Part 1: From Book to Vision',
  '2002'),
 (8.8, 'The Lord of the Rings: The Two Towers', '2002'),
 (7.9,
  "Creating the Lord of the Rings Symphony: A Composer's Journey Through Middle-Earth",
  '2004'),
 (7.8, "The Making of 'The Lord of the Rings'", '2002'),
 (7.6,
  'National Geographic: Beyond the Movie - The Lord of the Rings: The Fellowship of the Ring',
  '2001'),
 (7.6, "A Passage to Middle-earth: The Making of 'Lord of the Rings'", '2001'),
 (6.9, "Tolkien's the Lord of the Rings: A Catholic Worldview", '2011'),
 (6.6, 'The Lord of the Rings: The Quest Fulfilled', '2003'),
 (6.6,
  "Master of the Rings: The Unauthorized Story Behind J.R.R. Tolkien's 'Lord of the Rings'",
  '2001'),
 (6.3,
  'National Geographic: Beyond the Movie -

## Ejemplo con Go REST

Go REST (https://gorest.co.in/) es un API REST completa usada para hacer testing y prototipos. Su uso es gratutito y soporta las 4 operaciones CRUD. Para poder usarla es necesario darse de alta y obtener un _token_ de acceso que deberemos incluir en todas nuestras peticiones. Los datos creados/modificados por un usuario sólo son visibles para ese usuario en particular, y **todos los datos son borrados diariamente**. 

![Página Go REST](GoREST.png)


![Página Go REST](curl_examples.png)

In [10]:
# Pasos previos:
# 1) Crear un usuario en https://gorest.co.in/
# 2) Generar un token de acceso
# 3) Copia el token de acceso en el fichero 'token_gorest.txt' en el mismo directorio del notebook

with open('token_gorest.txt', 'r', encoding='utf8') as f:
    TOKEN_GOREST = f.read().strip()
    # print(TOKEN_GOREST)

In [11]:
from pprint import pprint, pp

# Función para mostrar los contenidos de la respuesta
def pprint_response(r):
    print(r.status_code)
    pp(dict(r.headers))
    print()
    pp(r.json())
    print(len(r.json()))

In [12]:
import requests

# Petición GET de todos los usuarios, API v2, token como parámetro y respuesta JSON
r = requests.get('https://gorest.co.in/public/v2/users', 
                 params={'access-token': TOKEN_GOREST})
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:48:55 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"bdbfaace5562dbebfc876a88cdcc27a0"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=1',
 'x-links-next': 'https://gorest.co.in/public/v2/users?page=2',
 'x-links-previous': '',
 'x-pagination-limit': '10',
 'x-pagination-page': '1',
 'x-pagination-pages': '291',
 'x-pagination-total': '2905',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '89',
 'x-ratelimit-reset': '1',
 'x-request-id': 'af758f53-9024-48bf-adb8-0e4c50c514ce',
 'x-runtime': '0.075522',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-

La cabecera de la contestación nos da información sobre los resultados obtenidos:
* `'x-links-current': 'https://gorest.co.in/public/v2/users?page=1'`: página actual
* `'x-links-next': 'https://gorest.co.in/public/v2/users?page=2'`: página siguiente
* `'x-links-previous': ''`: página siguiente, en este caso no hay
* `'x-pagination-page': '1'`: página actual
* `'x-pagination-limit': '10'`: valor por defecto para el tamaño de página
* `'x-pagination-pages': '291'`: número de páginas con el tamaño por defecto
* `'x-pagination-total': '2905'`: total de elementos de la colección    

In [13]:
import requests
from pprint import pprint

# Petición GET de todos los usuarios, API v1, token como parámetro y respuesta JSON
r = requests.get('https://gorest.co.in/public/v1/users', 
                 params={'access-token': TOKEN_GOREST})
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:48:56 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"531396964289ed78d9814b9180b80d01"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '89',
 'x-ratelimit-reset': '2',
 'x-request-id': '8fbc0c95-2408-4142-afda-47201b89dffc',
 'x-runtime': '0.120550',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=qCz0e5%2BeRMxKrLv%2BDhyt0DPZt%2BEoVVvCCG12uZoZ%2B2pr2PMCB87deglBSFyFXgOrWNxhYVT%2B9FR7w8id4XQIfYpvPQUGEev304q%2FGzBySElIEB95tMFA5aveEo3klBs%3D"}],"group":"cf-nel","max_age":604800}',
 'NEL': '{"su

In [14]:
from pprint import pprint

# Petición GET de todos los usuarios, API v2, token en la cabecera y respuesta JSON
r = requests.get('https://gorest.co.in/public/v2/users', 
                  headers={'Authorization': f'Bearer {TOKEN_GOREST}'})
# Cabeceras de la petición
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:48:56 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"bdbfaace5562dbebfc876a88cdcc27a0"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=1',
 'x-links-next': 'https://gorest.co.in/public/v2/users?page=2',
 'x-links-previous': '',
 'x-pagination-limit': '10',
 'x-pagination-page': '1',
 'x-pagination-pages': '291',
 'x-pagination-total': '2905',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '88',
 'x-ratelimit-reset': '2',
 'x-request-id': '0ec285e7-01ca-4e8c-a3a9-b52d4dee011a',
 'x-runtime': '0.136496',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-

In [15]:
import xml.dom.minidom

# Petición GET de todos los usuarios, API v2, token en la cabecera y respuesta XML
r = requests.get('https://gorest.co.in/public/v2/users.xml', 
                  headers={'Authorization': f'Bearer {TOKEN_GOREST}'})
pp(dict(r.headers))
print('-------------------------')
print(r.text)
# Si queremos que se muestre con más espacios en blanco
# dom = xml.dom.minidom.parseString(r.text)
# print(dom.toprettyxml())

{'Date': 'Mon, 14 Oct 2024 10:48:57 GMT',
 'Content-Type': 'application/xml; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"d56ae180c0acd197187cfb9faf80b309"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=1',
 'x-links-next': 'https://gorest.co.in/public/v2/users?page=2',
 'x-links-previous': '',
 'x-pagination-limit': '10',
 'x-pagination-page': '1',
 'x-pagination-pages': '291',
 'x-pagination-total': '2905',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '87',
 'x-ratelimit-reset': '2',
 'x-request-id': '884a5f47-3f8e-45c2-be99-dabc8dc71dfc',
 'x-runtime': '0.742775',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': 

In [16]:
from pprint import pprint

# Petición GET de todos los usuarios, API v2, token en la cabecera y respuesta JSON
# Paginación: página 2 contando de 10 en 10
r = requests.get('https://gorest.co.in/public/v2/users', 
                 params={'page': 2},
                 headers={'Authorization': f'Bearer {TOKEN_GOREST}'})
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:48:57 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"440e21aadbb9e6ac6a99c1c3c54f1b39"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=2',
 'x-links-next': 'https://gorest.co.in/public/v2/users?page=3',
 'x-links-previous': 'https://gorest.co.in/public/v2/users?page=1',
 'x-pagination-limit': '10',
 'x-pagination-page': '2',
 'x-pagination-pages': '291',
 'x-pagination-total': '2905',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '87',
 'x-ratelimit-reset': '3',
 'x-request-id': '0c99676c-ee80-4ccc-8763-b50a05693179',
 'x-runtime': '0.111387',
 'x-xss-protection': '0

In [17]:
from pprint import pprint

# Petición GET de todos los usuarios, API v2, token en la cabecera y respuesta JSON
# Paginación: página 4 contando de 3 en 3
r = requests.get('https://gorest.co.in/public/v2/users', 
                 params={'page': 4, 'per_page': 3},
                 headers={'Authorization': f'Bearer {TOKEN_GOREST}'}
                )
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:48:58 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"ed3c438f4c3363e0c1a7fa4d18d28247"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=4',
 'x-links-next': 'https://gorest.co.in/public/v2/users?page=5',
 'x-links-previous': 'https://gorest.co.in/public/v2/users?page=3',
 'x-pagination-limit': '3',
 'x-pagination-page': '4',
 'x-pagination-pages': '969',
 'x-pagination-total': '2905',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '86',
 'x-ratelimit-reset': '4',
 'x-request-id': '7c6e76fc-edfa-4218-90e6-c47d3d09b77c',
 'x-runtime': '0.152141',
 'x-xss-protection': '0'

In [18]:
# Acceder a una página que no existe devuelve la lista vacía pero no genera error
r = requests.get('https://gorest.co.in/public/v2/users', 
                 params={'page': 1000, 'per_page': 10},
                 headers={'Authorization': f'Bearer {TOKEN_GOREST}'})
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:48:58 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Content-Length': '2',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"4f53cda18c2baa0c0354bb5f9a3ecbe5"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=1000',
 'x-links-next': 'https://gorest.co.in/public/v2/users?page=1001',
 'x-links-previous': 'https://gorest.co.in/public/v2/users?page=999',
 'x-pagination-limit': '10',
 'x-pagination-page': '1000',
 'x-pagination-pages': '291',
 'x-pagination-total': '2905',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '85',
 'x-ratelimit-reset': '4',
 'x-request-id': '2adc4da1-b4af-43a0-ac47-81a89a96289f',
 'x-runtime': '0.065119',
 'x-xss-protection': 

In [19]:
# Introducir valores incorrectos en 'page' o 'per_page' no genera errores, únicamente cambia esos
# valores incorrectos por sus valores por defecto
r = requests.get('https://gorest.co.in/public/v2/users', 
                 params={'page': 'uno', 'per_page': 'siete'},
                 headers={'Authorization': f'Bearer {TOKEN_GOREST}'})
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:48:58 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"bdbfaace5562dbebfc876a88cdcc27a0"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=1',
 'x-links-next': 'https://gorest.co.in/public/v2/users?page=2',
 'x-links-previous': '',
 'x-pagination-limit': '10',
 'x-pagination-page': '1',
 'x-pagination-pages': '291',
 'x-pagination-total': '2905',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '84',
 'x-ratelimit-reset': '5',
 'x-request-id': 'f7202c7d-527f-484d-bd5d-341ee3b570d8',
 'x-runtime': '0.062501',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-

In [20]:
# Creación de un usuario
r = requests.post('https://gorest.co.in/public/v2/users',
                  headers={'Authorization': f'Bearer {TOKEN_GOREST}'},
                  data={'name': 'Facundo',
                        'email': 'facu@gmail.com',
                        'gender': 'male',
                        'status': 'active'}
                 )
pprint_response(r)
id_facundo = r.json()['id']
print(id_facundo)

201
{'Date': 'Mon, 14 Oct 2024 10:48:59 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Content-Length': '90',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"60a5bb8bba26e2515706d768ae144afa"',
 'location': 'https://gorest.co.in/public/v2/users/7469233',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '83',
 'x-ratelimit-reset': '6',
 'x-request-id': 'd5d5179c-d23e-4b81-9b80-2d6839138466',
 'x-runtime': '0.142814',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=50w9aSi6hGnyZld%2BTjXJH8PFA5w4HhRWy4%2FAN0f%2FvbHSCekMcHfVxQnpZf0Zke8apDKK6NBiZFkuJG7bCd8eolMDmViUhjrLN69%2BpJNOkxR8L5aB7CGCjSXHkjWN5jM%3D"}],"

In [21]:
# Búsqueda de usuario por email (usuarios que contengan la subcadena 'facu@gmail.com' en su email)
r = requests.get('https://gorest.co.in/public/v2/users', 
                 params={'email': 'facu@gmail.com'},
                 headers={'Authorization': f'Bearer {TOKEN_GOREST}'})
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:49:00 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"99e91bf290540d10baa0cad0cf5bb009"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=1',
 'x-links-next': '',
 'x-links-previous': '',
 'x-pagination-limit': '10',
 'x-pagination-page': '1',
 'x-pagination-pages': '1',
 'x-pagination-total': '1',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '82',
 'x-ratelimit-reset': '6',
 'x-request-id': '0714a89b-d271-4c77-bb12-71ef087dc653',
 'x-runtime': '0.628014',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cl

In [22]:
# Búsqueda de usuario por id
r = requests.get('https://gorest.co.in/public/v2/users', 
                 params={'id': id_facundo},
                 headers={'Authorization': f'Bearer {TOKEN_GOREST}'})
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:49:00 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"99e91bf290540d10baa0cad0cf5bb009"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-links-current': 'https://gorest.co.in/public/v2/users?page=1',
 'x-links-next': '',
 'x-links-previous': '',
 'x-pagination-limit': '10',
 'x-pagination-page': '1',
 'x-pagination-pages': '1',
 'x-pagination-total': '1',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '82',
 'x-ratelimit-reset': '6',
 'x-request-id': '61025f9a-84b9-4049-83c3-66bf186b74fb',
 'x-runtime': '0.057435',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cl

In [23]:
# Borrado de un usuario a partir del ID
r = requests.delete(f'https://gorest.co.in/public/v2/users/{id_facundo}',
                    headers={'Authorization': f'Bearer {TOKEN_GOREST}'}
)
print(r.status_code)
pp(dict(r.headers))
print(r.text)  # Importante: no devuelve ningún texto para pasar a JSON

204
{'Date': 'Mon, 14 Oct 2024 10:49:00 GMT',
 'Connection': 'keep-alive',
 'Cache-Control': 'no-cache',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '81',
 'x-ratelimit-reset': '7',
 'x-request-id': '4b70e283-8ed8-43a5-bc1e-a6d83aa0c4d6',
 'x-runtime': '0.097970',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=cm9QC2BdMYpWWivMp7Flly%2BBZaAURr3BnuMbNYi9YvuYW%2FDJEoFyLaaTvT3zjIsUD31gfzdft6Irr6y4mit%2FeHR96rDK2vJ6yFbIdOTwsfcua8vCPFUoSXNCsmkaI0M%3D"}],"group":"cf-nel","max_age":604800}',
 'NEL': '{"success_fraction":0,"report_to":"cf-nel","max_age":604800}',
 'Server': 'cloudflare',
 'CF-RAY': '8d2700125f80cfee-MAD',
 'alt-svc': 'h3=":443"; ma=86400'}



In [24]:
# Creación de un usuario para actualizarlo
r = requests.post('https://gorest.co.in/public/v2/users',
                  headers={'Authorization': f'Bearer {TOKEN_GOREST}'},
                  data={'name': 'Eva',
                        'email': 'eva@gmail.com',
                        'gender': 'female',
                        'status': 'inactive'}
                 )
pprint_response(r)
id_eva = r.json()['id']
print(id_eva)

201
{'Date': 'Mon, 14 Oct 2024 10:49:01 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Content-Length': '89',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"0c349414703433875ee0542150fdc329"',
 'location': 'https://gorest.co.in/public/v2/users/7469235',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '80',
 'x-ratelimit-reset': '8',
 'x-request-id': 'd78a87b2-8929-4d61-ae40-801246d0eb03',
 'x-runtime': '0.140471',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=5ZLqxehfeq8W2nEwUXY5f65B6j93B4As%2B4RCwj8N6fTDWfj4vqifrWG8OESxNPXTjRckMQTQlV6Kzkmgy0oxdWyeT%2FGIGhVL6sIwLRgysiBOZTx8UJZFc8YM6lV6Ne4%3D"}],"grou

In [25]:
# Modificación únicamente de un campo: PATCH
# (En GoREST PATCH permite modificaciones de varios campos)
r = requests.patch(f'https://gorest.co.in/public/v2/users/{id_eva}',
                  headers={'Authorization': f'Bearer {TOKEN_GOREST}'},
                  data={'status': 'active'}
                 )
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:49:01 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"30d105b84ea8e186dfb2efc7e442a418"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '79',
 'x-ratelimit-reset': '8',
 'x-request-id': 'b449f5da-f2ce-464b-868e-a54f073dcb3f',
 'x-runtime': '0.111139',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=OzBMfCR1WYVAhyQZetWG73AhBogFwtU7a8tDXsJouGofbSAauC%2FWC2dKtQiCOFaIxBPeBQXMKJqPh32p9Fk2gmekN00d1CNShDmOVULiVyw6WUFlw531u2KXBxYDlRg%3D"}],"group":"cf-nel","max_age":604800}',
 'NEL': '{"success_frac

In [26]:
# Modificación de todos los campos menos el identificador: PUT
r = requests.put(f'https://gorest.co.in/public/v2/users/{id_eva}',
                  headers={'Authorization': f'Bearer {TOKEN_GOREST}'},
                  data={'name': 'Eva María',
                        'email': 'evam@gmail.com',
                        'gender': 'female',
                        'status': 'inactive'}
                 )
pprint_response(r)

200
{'Date': 'Mon, 14 Oct 2024 10:49:01 GMT',
 'Content-Type': 'application/json; charset=utf-8',
 'Transfer-Encoding': 'chunked',
 'Connection': 'keep-alive',
 'Cache-Control': 'max-age=0, private, must-revalidate',
 'etag': 'W/"5cbacbc4ab9c76e9d1302751710d5074"',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '78',
 'x-ratelimit-reset': '9',
 'x-request-id': '17ad124b-2c27-42f4-9d86-449c68a2062c',
 'x-runtime': '0.071392',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=Ic5cHds3GzR8pPG%2B0x%2BncoycZnl6%2BOtJ439OIF7d6z36NTTwAYQwnNCUbKY%2BLRr5jC1jp%2Fsd8P7YLmglZ3oPKThYr1igoXnActMYI4oRZ1uq42ZF%2FPwGKVxwQXtlAyk%3D"}],"group":"cf-nel","max_age":604800}',
 'NEL': '{"su

In [27]:
# Por limpieza borramos también a EVA
r = requests.delete(f'https://gorest.co.in/public/v2/users/{id_eva}',
                 headers={'Authorization': f'Bearer {TOKEN_GOREST}'})
print(r.status_code)
pp(dict(r.headers))
print(r.text)

204
{'Date': 'Mon, 14 Oct 2024 10:49:02 GMT',
 'Connection': 'keep-alive',
 'Cache-Control': 'no-cache',
 'referrer-policy': 'strict-origin-when-cross-origin',
 'vary': 'Origin',
 'x-content-type-options': 'nosniff',
 'x-download-options': 'noopen',
 'x-frame-options': 'SAMEORIGIN',
 'x-permitted-cross-domain-policies': 'none',
 'x-ratelimit-limit': '90',
 'x-ratelimit-remaining': '77',
 'x-ratelimit-reset': '10',
 'x-request-id': '3dbf4b8b-4c8d-4681-a7c4-c57b0370c532',
 'x-runtime': '0.118420',
 'x-xss-protection': '0',
 'cf-cache-status': 'DYNAMIC',
 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=0YxE6ITAe1DN%2B2NeMfFd0qSkCkVpJGnGjWqR%2FyKmp9ghRPGpvcOimsV3Q4VmoHStpnY%2FT%2FcTh%2FyYeGlgMQZ%2BC%2FajOWmp9jwA7N7KEgE7rDj2IE2Rh38YWW2FZC26lTk%3D"}],"group":"cf-nel","max_age":604800}',
 'NEL': '{"success_fraction":0,"report_to":"cf-nel","max_age":604800}',
 'Server': 'cloudflare',
 'CF-RAY': '8d27001b3efccbc8-MAD',
 'alt-svc': 'h3=":443"; ma=86400'}

